File: sdk/java/src/main/java/io/forma/CtxPrimitive.java
package io.forma;

import java.util.*;

/**
 * One ctx primitive handle (db/cache/lock/…), optionally bound to a named
 * datastore via {@link #named(String)}. Operations map 1:1 to
 * {@code POST /ctx/{prim}/{op}}.
 */
public final class CtxPrimitive {
    private final SidecarClient client;
    private final String type;
    private final String boundName;

    CtxPrimitive(SidecarClient client, String type) {
        this(client, type, "");
    }

    CtxPrimitive(SidecarClient client, String type, String boundName) {
        this.client = client;
        this.type = type;
        this.boundName = boundName;
    }

    /** Bind to a named datastore instead of the default one. */
    public CtxPrimitive named(String name) {
        return new CtxPrimitive(client, type, name);
    }

    // ---- db ----

    /**
     * Execute a query on the datastore.
     *
     * @param sql  the SQL query string
     * @param args positional arguments (optional)
     * @return list of result rows as maps
     */
    @SuppressWarnings("unchecked")
    public List<Map<String, Object>> query(String sql, Object... args) {
        var body = new LinkedHashMap<String, Object>();
        body.put("sql", sql);
        if (args.length > 0) {
            body.put("args", List.of(args));
        }
        var resp = call("query", body);
        var data = resp.get("data");
        return data instanceof List ? (List<Map<String, Object>>) data : List.of();
    }

    // ---- cache / kvstore ----

    /**
     * Get a value by key.
     *
     * @return the value, or {@code null} if not found
     */
    public Object get(String key) {
        return call("get", Map.of("key", key)).get("data");
    }

    /**
     * Set a value by key.
     *
     * @param ttlSeconds time-to-live in seconds (0 = no expiry)
     */
    public void set(String key, Object value, int ttlSeconds) {
        var body = new LinkedHashMap<String, Object>();
        body.put("key", key);
        body.put("value", value);
        if (ttlSeconds > 0) {
            body.put("ttl_seconds", ttlSeconds);
        }
        call("set", body);
    }

    /** Delete a key. */
    public void delete(String key) {
        call("delete", Map.of("key", key));
    }

    // ---- lock ----

    /**
     * Acquire a distributed lock.
     *
     * @param key        the lock key
     * @param ttlSeconds lock TTL in seconds (default 30)
     * @return {@code true} if the lock was acquired
     */
    public boolean acquire(String key, int ttlSeconds) {
        var resp = call("acquire", Map.of("key", key, "ttl_seconds", ttlSeconds));
        return Boolean.TRUE.equals(resp.get("ok"));
    }

    /** Release a distributed lock. */
    public void release(String key) {
        call("release", Map.of("key", key));
    }

    // ---- entity atomic operations ----

    /**
     * Atomically merge fields into an entity record (entity/update).
     * Uses jsonb_merge / json_patch — single SQL statement, no race condition.
     * Workspace isolation is enforced by the sidecar — not a parameter.
     */
    public void update(String id, Map<String, Object> fields) {
        var body = new LinkedHashMap<String, Object>();
        body.put("key", id);
        body.put("fields", fields);
        call("update", body);
    }

    /**
     * Atomically increment a numeric field on an entity record.
     * Single SQL statement — no read-modify-write race condition.
     * Workspace isolation is enforced by the sidecar — not a parameter.
     */
    public void increment(String id, String field, double amount) {
        var body = new LinkedHashMap<String, Object>();
        body.put("key", id);
        body.put("field", field);
        body.put("amount", amount);
        call("increment", body);
    }

    /**
     * Atomically decrement a numeric field on an entity record.
     * Includes a guard against negative values. Returns the new field value.
     * Workspace isolation is enforced by the sidecar — not a parameter.
     */
    public Object decrement(String id, String field, double amount) {
        var body = new LinkedHashMap<String, Object>();
        body.put("key", id);
        body.put("field", field);
        body.put("amount", amount);
        var resp = call("decrement", body);
        return resp.get("data");
    }

    // ---- internal ----

    private Map<String, Object> call(String op, Map<String, Object> body) {
        if (!boundName.isEmpty()) {
            body = new LinkedHashMap<>(body);
            body.put("named", boundName);
        }
        return client.post("/ctx/" + type + "/" + op, body);
    }
}

File: sdk/php/src/CtxPrimitive.php
<?php

declare(strict_types=1);

namespace Forma;

/**
 * One ctx primitive handle (db/cache/lock/...), optionally bound to a named
 * datastore via named(). Operations map 1:1 to POST /ctx/{prim}/{op}.
 */
final class CtxPrimitive
{
    public function __construct(
        private readonly SidecarClient $client,
        private readonly string $type,
        private readonly string $named = '',
    ) {
    }

    /** Bind to a named datastore instead of the default one. */
    public function named(string $name): self
    {
        return new self($this->client, $this->type, $name);
    }

    /**
     * @param list<mixed> $args
     * @return list<array<string, mixed>> result rows
     */
    public function query(string $sql, array $args = []): array
    {
        $body = ['sql' => $sql];
        if ($args !== []) {
            $body['args'] = $args;
        }

        return (array) ($this->call('query', $body)['data'] ?? []);
    }

    public function get(string $key): mixed
    {
        return $this->call('get', ['key' => $key])['data'] ?? null;
    }

    public function set(string $key, mixed $value, int $ttlSeconds = 0): void
    {
        $body = ['key' => $key, 'value' => $value];
        if ($ttlSeconds > 0) {
            $body['ttl_seconds'] = $ttlSeconds;
        }
        $this->call('set', $body);
    }

    public function delete(string $key): void
    {
        $this->call('delete', ['key' => $key]);
    }

    // ---- entity atomic operations ----

    /**
     * Atomically merge fields into an entity record (entity/update).
     * Uses jsonb_merge / json_patch — single SQL statement, no race condition.
     */
    public function update(string $id, array $fields): void
    {
        $body = ['key' => $id, 'fields' => $fields];
        $this->call('update', $body);
    }

    /**
     * Atomically increment a numeric field on an entity record.
     * Single SQL statement — no read-modify-write race condition.
     */
    public function increment(string $id, string $field, float $amount): void
    {
        $body = ['key' => $id, 'field' => $field, 'amount' => $amount];
        $this->call('increment', $body);
    }

    /**
     * Atomically decrement a numeric field on an entity record.
     * Includes a guard against negative values. Returns the new field value.
     */
    public function decrement(string $id, string $field, float $amount): mixed
    {
        $body = ['key' => $id, 'field' => $field, 'amount' => $amount];
        return $this->call('decrement', $body)['data'] ?? null;
    }

    /** @return bool true if the lock was acquired */
    public function acquire(string $key, int $ttlSeconds = 30): bool
    {
        return (bool) ($this->call('acquire', ['key' => $key, 'ttl_seconds' => $ttlSeconds])['ok'] ?? false);
    }

    public function release(string $key): void
    {
        $this->call('release', ['key' => $key]);
    }

    /**
     * @param array<string, mixed> $body
     * @return array<string, mixed>
     */
    private function call(string $op, array $body): array
    {
        if ($this->named !== '') {
            $body['named'] = $this->named;
        }

        return $this->client->post("/ctx/{$this->type}/{$op}", $body);
    }
}
